coalib/nestedlib/NlSection.py

```diff
@@ -0,0 +1,122 @@
+                            than start as negative ranges are not allowed.
+        :param language:    The programming language of the lines.
+        :param index:       The index of the nl_section.
+        :param file:        The name of the original file
+        :param temp_file:   The name of the temporary segregated file
+        :raises TypeError:  Raised when
+                            - start is not of type NlSectionPosition.
+                            - end is neither of type NlSectionPosition, nor
+                              is it None.
+                            - file is not of type str
+                            - index is not of type int
+                            - language is not of type str
+        :raises ValueError: Raised when file of start and end mismatch.
+        """
+        TextRange.__init__(self, start, end)
+        self.index = index
+        self.language = language
+
+        """
+        :linted_start: The start of the section in the linted file.Initially it
+                       is same as that of the start of the original file. It
+                       changes only when any patches are applied on that line.
+        :linted_end:   The end of the section in the linted file.Initially it
+                       is same as that of the end of the original file. It
+                       changes only when any patches are applied on that line.
+                       If ``None`` is given, the start object will be used
+                       here
+        """
+        self.linted_start = NlSectionPosition(file, start.line, start.column)
+        if end:
+            self.linted_end = NlSectionPosition(file, end.line, end.column)
+        else:
+            self.linted_end = NlSectionPosition(file, start.line, start.column)
+
+        if self.start.file != self.end.file:
+            raise ValueError('File of start and end position do not match.')
+
+    @classmethod
+    def from_values(cls,
+                    file,
+                    index,
+                    language,
+                    start_line=None,
+                    start_column=None,
+                    end_line=None,
+                    end_column=None,):
+        start = NlSectionPosition(file, start_line, start_column)
+        if end_line or (end_column and end_column > start_column):
+            end = NlSectionPosition(file, end_line if end_line else start_line,
+                                    end_column)
+        else:
+            end = None
+        return cls(file, index, language, start, end)
+
+    @property
+    def file(self):
+        return self.start.file
+
+    def __str__(self):
+        """
+        Create a string representation of the NlSection object.
+
+        If the whole file is affected, then just the filename is shown.
+
+        >>> str(NlSection.from_values('test_file', 1, 'html', None, None,\
+                                      None, None))
+        '...test_file: 1: html'
```

Comment:
Doctests fail here, note that your expected string doesn't have an extra `: ` at the end whereas that's what the function is returning. I think this is a bug? (and a case not covered in your unittests)

Reply:
Also inconsistent spacing at some places in the docstrings after Line and Column specifications

Reply:
Ah! snap. This was the old doctest. I forgot to update it when I made the changes. Would change them now.